File: 백준/Gold/17298. 오큰수/오큰수.java
import java.io.*;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] nums = new int[n];
        int[] results = new int[n];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }
        br.close();

        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(0);
        for (int i = 1; i < n; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] < nums[i]) {
                results[stack.pop()] = nums[i];
            }
            stack.push(i);
        }
        
        while (!stack.isEmpty()) results[stack.pop()] = -1;

        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        for (int i = 0; i < n; i++) {
            bw.write(results[i] + " ");
        }
        
        bw.write("\n");
        bw.flush();
        bw.close();
    }
}

File: 백준/Gold/17298. 오큰수/오큰수.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val nums = IntArray(n)
    val results = IntArray(n)

    val st = StringTokenizer(br.readLine())
    for (i in 0 until n) nums[i] = st.nextToken().toInt()
    br.close()

    val stack = ArrayDeque<Int>()
    stack.addLast(0)

    for (i in 1 until n) {
        while (stack.isNotEmpty() && nums[stack.last()] < nums[i]) {
            results[stack.removeLast()] = nums[i]
        }
        stack.addLast(i)
    }

    while (stack.isNotEmpty()) results[stack.removeLast()] = -1

    val bw = BufferedWriter(OutputStreamWriter(System.out))
    for (i in 0 until n) bw.write("${results[i]} ")
    bw.write("\n")
    bw.flush()
    bw.close()
}